### ISSUE TITLE
march=atom compiletime issues

### ISSUE THREAD

**Issue by arndb** [ClangBuiltLinux/linux#1483]
Building x86 with CONFIG_MATOM has multiple problems on clang-14 and earlier versions:

make[4]: *** [/git/arm-soc/scripts/Makefile.build:288: arch/x86/hyperv/hv_init.o] Error 1
error: ran out of registers during register allocation
error: ran out of registers during register allocation

make[6]: *** [/git/arm-soc/scripts/Makefile.build:288: drivers/gpu/drm/vmwgfx/vmwgfx_msg.o] Error 1
error: ran out of registers during register allocation
error: ran out of registers during register allocation
2 errors generated.

arch/x86/crypto/curve25519-x86_64.c:610:3: error: inline assembly requires more registers than available

and hundreds of:
drivers/hwtracing/stm/core.o: warning: objtool: stm_char_write()+0x18c: unreachable instruction
drivers/android/binder.o: warning: objtool: binder_transaction()+0x1aed: unreachable instruction
drivers/hwtracing/intel_th/gth.o: warning: objtool: intel_th_gth_enable()+0x41: unreachable instruction
drivers/hwtracing/intel_th/msu.o: warning: objtool: intel_th_msc_probe()+0x120: unreachable instruction
drivers/mux/core.o: warning: objtool: mux_chip_alloc()+0x1d4: unreachable instruction
drivers/interconnect/core.o: warning: objtool: icc_set_bw()+0x2a: unreachable instruction
drivers/rapidio/devices/tsi721.o: warning: objtool: tsi721_rio_map_inb_mem()+0x47a: unreachable instruction
drivers/dax/bus.o: warning: objtool: size_store()+0x635: unreachable instruction
drivers/input/serio/libps2.o: warning: objtool: ps2_drain()+0x2e: unreachable instruction
drivers/greybus/core.o: warning: objtool: greybus_uevent()+0x86: unreachable instruction
drivers/hid/usbhid/hiddev.o: warning: objtool: hiddev_ioctl_string()+0xae: unreachable instruction
drivers/bcma/driver_chipcommon.o: warning: objtool: bcma_chipco_watchdog_timer_set()+0x4c: unreachable instruction
drivers/input/serio/userio.o: warning: objtool: userio_char_read()+0x256: unreachable instruction
drivers/comedi/comedi_fops.o: warning: objtool: comedi_read()+0x2d8: unreachable instruction
drivers/ntb/hw/epf/ntb_hw_epf.o: warning: objtool: ntb_epf_send_command()+0x9b: unreachable instruction

These were hidden from my randconfig builds as I had disabled the option to work around https://llvm.org/pr42565. However, that bug is now fixed, and enabling MATOM again now runs into the other issues.

**Comment by arndb** [comment]
Attaching four randconfig files and their output:
[config-matom.tar.gz](https://github.com/ClangBuiltLinux/linux/files/7371360/config-matom.tar.gz)

**Comment by arndb** [comment]
Submitted a workaround: https://lore.kernel.org/all/20211019171121.3510624-1-arnd@kernel.org/T/#u

**Comment by nickdesaulniers** [comment]
Easily observable with defconfig + CONFIG_MATOM=y.

I thought this might be related to #1566, but the patch I have for that doesn't change anything here AFAICT.

A reduction produced:
```c
// clang -O2 -march=atom foo.c
__register_chrdev() { asm(".byte 15, 0x0b"); }
```
```
char_dev.o: warning: objtool: __register_chrdev()+0x8: unreachable instruction
```
```sh
$ llvm-objdump -dr char_dev.o

char_dev.o:     file format elf64-x86-64

Disassembly of section .text:

0000000000000000 <__register_chrdev>:
       0: 0f 0b                         ud2
       2: 90                            nop
       3: 90                            nop
       4: 90                            nop
       5: 90                            nop
       6: 90                            nop
       7: 90                            nop
       8: c3                            retq
```

Not sure about that though. That must have been over-reduced because `asm(".byte 15, 0x0b");` is adding the trapping `ud2` which makes the `retq` at `__register_chrdev()+0x8` unreachable.

**Comment by nickdesaulniers** [comment]
> I thought this might be related to #1566, but the patch I have for that doesn't change anything here AFAICT.

Hmm...extracting the function in IR, I do see `.section        .discard.reachable,"",@progbits`, so maybe it is related.

**Comment by nickdesaulniers** [comment]
```diff
diff --git a/arch/x86/include/asm/bug.h b/arch/x86/include/asm/bug.h
index 84b87538a15d..20201f1aedeb 100644
--- a/arch/x86/include/asm/bug.h
+++ b/arch/x86/include/asm/bug.h
@@ -76,7 +76,6 @@ do {                                                          \
 #define __WARN_FLAGS(flags)                                    \
 do {                                                           \
        instrumentation_begin();                                \
-       _BUG_FLAGS(ASM_UD2, BUGFLAG_WARNING|(flags));           \
        annotate_reachable();                                   \
        instrumentation_end();                                  \
 } while (0)
```
is enough to eliminate all of these but one.  (As does deleting `HAVE_ARCH_BUG`, `BUG`, and `__WARN_FLAGS` from that same file). I suspect that the `annotate_reachable()`/`annotate_unreachable()`/`instrumentation_begin()`/`instrumentation_end()` stuff is subtly broken.  It assumes that instructions can't be reordered across an asm statement, and thus that a label created in inline asm will always refer to all of the instructions from the following C statement, which simply isn't true.  If we used `asm goto` somehow and passed in the address of a label, then the compiler would know that the statement pointed to by the label was an important control flow barrier.

**Comment by nickdesaulniers** [comment]
> is enough to eliminate all of these but one.

Which is #1563. Heh, I'm getting a sinking feeling that #1566, #1563, and #1483 are all interrelated.

**Comment by nickdesaulniers** [comment]
FWIW: https://godbolt.org/z/cfsc4vdGK

I have [a patch](https://github.com/ClangBuiltLinux/linux/commit/473c4f04fa49864f6955062435d8bab02ff87862), but testing is tricky since there are so many related objtool warnings.

**Comment by nickdesaulniers** [comment]
submitted: https://lore.kernel.org/lkml/20220202205557.2260694-1-ndesaulniers@google.com/

**Comment by nickdesaulniers** [comment]
accepted: https://git.kernel.org/pub/scm/linux/kernel/git/tip/tip.git/commit/?h=objtool/urgent&id=bfb1a7c91fb7758273b4a8d735313d9cc388b502

**Comment by nickdesaulniers** [comment]
commit bfb1a7c91fb7 ("x86/bug: Merge annotate_reachable() into _BUG_FLAGS() asm")
available in v5.17-rc4.

I did a scrub in the upstream llvm issue tracker for `ran out of registers during register allocation`. It looks like there are quite a few open issues specific to i386:
1. https://github.com/llvm/llvm-project/issues/41914
2. https://github.com/llvm/llvm-project/issues/11767
3. https://github.com/llvm/llvm-project/issues/16170

---

Trying to reproduce the reported error in arch/x86/hyperv/hv_init.o.
I tried ARCH=i386 defconfig plus
HYPERVISOR_GUEST=y
CONFIG_HYPERV=y
CONFIG_MATOM=y

but was unable to reproduce on mainline w/ clang-15.

---

Trying to reproduce the reported error in drivers/gpu/drm/vmwgfx/vmwgfx_msg.o.

I tried ARCH=i386 defconfig plus
CONFIG_DRM_VMWGFX=y
CONFIG_MATOM=y

but was unable to reproduce on mainline w/ clang-15.

---

Trying to reproduce the failure in arch/x86/crypto/curve25519-x86_64.c

I tried ARCH=i386 defconfig plus
CONFIG_WIREGUARD=y
CONFIG_CRYPTO_CURVE25519_X86=y
CONFIG_MATOM=y

but CONFIG_CRYPTO_CURVE25519_X86 is not selectable if !`64BIT`.

---

Since there are open issues upstream, I'm not convinced this is the last we'll see of register exhaustion related errors on i386 (`ran out of registers during register allocation` or `inline assembly requires more registers than available`).

But this bug is a mess of multiple different reports for different things, so closing this out since the objtool warnings are fixed by 

commit bfb1a7c91fb7 ("x86/bug: Merge annotate_reachable() into _BUG_FLAGS() asm")
available in v5.17-rc4.

I assume the rest of the issues are coming from randconfigs. Let's get bugs filed for individual cases when we find them again. It would be helpful to isolate which configs are necessary to reproduce from defconfig.
